Issue/augment key not mandatory

gmuloc opened the pull request:
Issue:

When using the tree plugin the keys of a list inside an augment are being marked as optional (the parent model is in the same folder):

```
yanger -f tree small6.yang
```
output (notice test-key is marked as optional):
```                                                                                                                                                   
module: small6

  augment /parent:parent-container:
    +--rw test-list* [test-key]
       +--rw test-key?    string
       +--rw dummy?      uint16
```

For comparison here is pyang output for the same model:
```
pyang -f tree small6.yang                                                                                                                                                      
```
output:
```

module: small6
  augment /parent:parent-container:
    +--rw test-list* [test-key]
       +--rw test-key    string
       +--rw dummy?      uint16
```

This is due to the following line in the yang tree plugin:
```
             NewName =
                 case
                     PKey /= undefined
                     andalso ?in(Sn#sn.name, PKey)
```

the andalso statement will always be false inside an augment as the name looks like `{_, name}`

I have integrated a fix in this PR that makes the logic work and I have added a test - however I am getting an issue with the way pyang and yanger tree plugins are inserting new lines in this scenario.


**Remaining problem on which I need help:** The output from pyang and yanger is different in this case because of the newline inserted before `module` in pyang and after `module` in yanger

gmuloche committed:
Added way to verify if an augmented leaf is a key in tree plugin

* Also fix some newline behavior for different sections to follow
  RFC 8340 section 2.

ksallberg commented:
For your question, please see line ~139:

    SectionPrinted1 = true, % always print separator after data, even if empty
    maybe_print_separator(Fd, SectionPrinted1 andalso Augs /= []),

'SectionPrinted1=true' will cause maybe_print_separator to print a newline before an augmented node. I don't know if this is good or bad really, it could be changed to not do this obviously. But IMO it doesn't hurt too much having a blank line there either. Is it causing trouble with your internal testing or something?

Edit: Anyway, I think SectionPrinted1 should probably be removed. And possibly this call to maybe_print_separator should be removed too which would cause the newline to go away

perhedeland commented:
Just in case you don't already know, there is actually an RFC specifying the tree format these days: https://tools.ietf.org/html/rfc8340 . I can't say that it provides a definitive answer to the question here, but at least looking at https://tools.ietf.org/html/rfc8340#section-2 , it seems that each "section" except # 1 ("The top-level data nodes defined in the module") should be preceded by an empty line. I.e. my vote would be that yanger is correct and pyang (based on the description here, I haven't tried a recent version) is wrong in this case.

ksallberg commented:
Per that RFC had escaped me, thanks for telling us. Gmuloche can you please remove that SectionPrinted1 printed variable, change the boolean expression and squash everything to one commit?

gmuloc commented:
Apologies I was on vacation the past couple of weeks and did not have time to look at this - will complete this this week

gmuloc commented:
@ksallberg - I had to do a bit more regarding the SectionPrinted variable - essentially because when I removed the SectionPrinted1 and changed the boolean expression it was breaking the sx.lux test because no new line was printed before the structure (while structure is not mentioned in RFC 8340 because it is still a draft, the draft indicates that it is supposed to replace yang-data defined in RFC 8040 so I assume it should receive the same treatment regarding a new line).

Regarding my test - I created a separate test for the augment just like the sx.lux one and added a comment in the main tree_plugin_macro.lux pointing to the RFC and this issue.

Probably a next step would be to open a PR in pyang as well.

ksallberg commented:
I initially didn't look carefully enough to see that SectionPrinted1 was used later on (in SectionPrinted2), so my earlier suggestion was a bit incomplete, sorry about that.

Now after your latest change, you changed the following (where SectionPrinted2 was 'Augs /= []')

`maybe_print_separator(Fd, SectionPrinted2 andalso Rpcs /= []),`

to

`maybe_print_separator(Fd, Rpcs /= []),`

This changes the logic a bit right, so that the separator is printed even if Augs is not empty. (And similarly, all following places where you remove SectionPrintedN are also different now) I am not sure if that is correct behaviour or not? I am thinking the new code might print several newlines on a row in some cases.

Would all the test cases (even the one you added) still pass without the changes you propose?

gmuloc commented:
In this expression

```
maybe_print_separator(Fd, SectionPrinted2 andalso Rpcs /= []),
```

SectionPrinted2 was originally `Augs /= [] orelse SectionPrinted1` and `SectionPrinted1` was always true (the line I removed) - so `SectionPrinted2` was thus always true and so `SectionPrintedN = Foo/=[] orelse SectionPrinted(N-1)` was always true no matter what.

The `maybe_print_separator` calls were always checking `SectionPrintedN andalso Foo /= []`  and as `SectionPrintedN` was always true (previous paragraph) the boolean expression was logically equivalent to `True andalso Foo /= []` so in the end it was logically equivalent to `Foo /= []`.

So my change does not really change the logic I believe.

Without all the changes I propose the tests I wrote would pass if I keep `SectionPrinted1 = true` however the logic is more circumvolved as these variables are essentially always true in the context and so not really useful. The reason why my test was not passing is because of the fact that pyang does not introduce the newline before the Augment when according to RFC it should as per @ksallberg  comment. This has not changed and that's why I moved the test to a separate lux file for now.

The reason why the sx tests started failing when I removed the `SectionPrinted1` variabl is that `SectionPrinted2` became `Augs /= []` which was `False` for the sx.yang file and so all the `SectionPrintedN` became `False` (except SectionPrinted5 but this is not used and anyway it is because of `Structs /= []`) thus the boolean expression `SectionPrinted4 andalso Structs /= []` was `False` and yanger ended printing no new line between the module line and the structure line:

```
    module: sx
      structure address-book:
        +-- address* [last first]
           +-- last      string
           +-- first     string
           +-- street?   string
           +-- city?     string
           +-- state?    string
```

rather than the (extrapolation of) the RFC (for structure) expected:
```
    module: sx

      structure address-book:
        +-- address* [last first]
           +-- last      string
           +-- first     string
           +-- street?   string
           +-- city?     string
           +-- state?    string
```

thus failing the existing test.

ksallberg commented:
Oh right, it was 'orelse' all the way from an initial true value... Then I think your PR is an improvement @gmuloc 👍

gmuloc commented:
Hello - is there anything else that I need to do to have this PR merged? just checking in case I missed something

mbj4668 commented:
I had a look at this and it looks good.  Thanks for the fix!  I will also push a fix to pyang for the newline issue, and update the test case here.
